cpp/CMakeLists.txt

```diff
@@ -28,6 +28,12 @@ project(
   VERSION 22.08.00
   LANGUAGES C CXX CUDA
 )
+if(CMAKE_CUDA_COMPILER_ID STREQUAL "NVIDIA" AND CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11.5)
+  message(
+    FATAL_ERROR
+      "libcudf requires CUDA Toolkit 11.5+ to compile (nvcc ${CMAKE_CUDA_COMPILER_VERSION} provided)"
+  )
```

Comment:
This makes my "single source of truth" sense tingle. This is the kind of thing that tends to get stale and out of date. Can this be derived from some other source of truth?

Reply:
I can't think of anyway we can pull this information from a reliable source of truth. 

While currently our conda dev env pins to 11.5, this could change in the future. For example libcudf still requires at least 11.5 but our python bindings due to a bug require 11.7+.

Reply:
> I can't think of anyway we can pull this information from a reliable source of truth.

Agreed. This is already hardcoded in several documents including the build requirements in CONTRIBUTING.md and conda environments, among others, but I don’t know of a better way to define it.

> For example libcudf still requires at least 11.5 but our python bindings due to a bug require 11.7+.

The Python bindings build with CTK 11.5 and newer. Are you referring to the supported versions of the cuda-python package? [Currently `>=11.5,<11.7.1`](https://github.com/rapidsai/cudf/blob/e98feab966f6d1b9eba83323e851c955a1691865/conda/environments/cudf_dev_cuda11.5.yml#L41) (version 11.7.1 pins an unsupported cuda-toolkit instead of the supported cudatoolkit, see [RSN 0020](https://docs.rapids.ai/notices/rsn0020/)).

Reply:
> The Python bindings build with CUDA 11.5 and newer. Are you referring to the supported versions of the cuda-python package?

I meant that in the future the required minimum CUDA versions for the C++ and Python sides could diverge and by using the conda package version we wouldn't properly capture that fact.

Reply:
Ah. Didn’t realize that was a hypothetical. 😅

Reply:
I also don't see an easy way around the duplication here, and I think the improved error handling is much friendlier and worth adding. @jrhemstad unless you object I'm inclined to go ahead and merge this.

Reply:
Once we resolve this conversation or get one more approval from @vyasr or @jrhemstad, I think this is ready to merge.

Reply:
Just a question. Can we do we something like we do when we check for the compiler and create a test build with a small generated source file that contains an int128 reference? So if that test build fails we know we have in incompatible compiler. We can use that to add features we need to check in the future as well. This way it is not checking the exact version as much as what it supports.

Reply:
I think that we could do that, but then I guess we'd be testing something slightly different than what we document our library as requiring. I don't know if there's any real risk there of things getting out of alignment, though.

Given that this open discussion has generated further questions since @bdice asked for another review, and since @robertmaynard has already moved this to the 22.10 board, I think we can afford to take a couple more days to consider alternatives, but realistically I suspect that our best option will just be merging this as is.

Reply:
> Can we do we something like we do when we check for the compiler and create a test build with a small generated source file that contains an int128 reference? So if that test build fails we know we have in incompatible compiler. 

This approach can be a  trap. What we have confirmed is that the compiler supports at maximum compiling the provided test case. It presumes that both the test case capture all our needs, and that we don't have issues with code generation correctness or performance that require a newer compiler version.

Reply:
That’s a good argument @robertmaynard. Let’s get a second approval here and merge.